File: javascript/core/src/verbs/unroll.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */
import type { TableStep } from './nodeFactories/index.js'
import type { InputColumnListArgs } from './types.js'
import { stepNodeFactory } from './nodeFactories/StepNode.js'

export type UnrollArgs = InputColumnListArgs

export const unrollStep: TableStep<UnrollArgs> = (input, { columns }) =>
	input.unroll(columns)

export const unroll = stepNodeFactory(unrollStep)

File: python/datashaper/datashaper/verbs/unroll.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#
"""Unroll verb implementation."""

import pandas as pd
from reactivedataflow import ConfigPort, InputPort, verb

from datashaper import DEFAULT_INPUT_NAME

from .decorators import OutputMode, wrap_verb_result


@verb(
    name="unroll",
    ports=[
        InputPort(name=DEFAULT_INPUT_NAME, parameter="table", required=True),
        ConfigPort(name="column", required=True),
    ],
    adapters=[
        wrap_verb_result(mode=OutputMode.Table),
    ],
)
def unroll(table: pd.DataFrame, column: str) -> pd.DataFrame:
    """Unroll a column."""
    return table.explode(column).reset_index(drop=True)
